tests/functional/directives.ts

```diff
@@ -0,0 +1,655 @@
+// tslint:disable:member-ordering
+import "reflect-metadata";
+
+import {
+  GraphQLSchema,
+  FieldDefinitionNode,
+  GraphQLField,
+  GraphQLDirective,
+  DirectiveLocation,
+  GraphQLString,
+  parseValue,
+  graphql,
+  GraphQLInputObjectType,
+  InputValueDefinitionNode,
+  InputObjectTypeDefinitionNode,
+  GraphQLInputField,
+  GraphQLScalarType,
+  GraphQLNonNull,
+} from "graphql";
+import {
+  Field,
+  InputType,
+  Resolver,
+  Query,
+  Arg,
+  Directive,
+  buildSchema,
+  ObjectType,
+  Mutation,
+} from "../../src";
+import { getMetadataStorage } from "../../src/metadata/getMetadataStorage";
+import Maybe from "graphql/tsutils/Maybe";
+import { SchemaDirectiveVisitor } from "graphql-tools";
+import { ApolloServer } from "apollo-server";
+
+class UpperCaseType extends GraphQLScalarType {
+  constructor(type: any) {
+    super({
+      name: "UpperCase",
+      parseValue: value => type.parseValue(value),
+      serialize: value => type.serialize(value),
+      parseLiteral: ast => {
+        const result = type.parseLiteral(ast);
+
+        if (typeof result === "string") {
+          return result.toUpperCase();
+        }
+
+        return result;
+      },
+    });
+  }
+}
+
+class UpperCaseDirective extends SchemaDirectiveVisitor {
+  visitFieldDefinition(field: GraphQLField<any, any>) {
+    const resolve = field.resolve;
+
+    field.resolve = async function(...args) {
+      const result = await resolve!.apply(this, args);
+      if (typeof result === "string") {
+        return result.toUpperCase();
+      }
+
+      return result;
+    };
+  }
+
+  visitInputObject(object: GraphQLInputObjectType) {
+    const fields = object.getFields();
+
+    Object.keys(fields).forEach(field => {
+      this.visitInputFieldDefinition(fields[field]);
+    });
+  }
+
+  visitInputFieldDefinition(field: GraphQLInputField) {
+    if (field.type instanceof UpperCaseType) {
+      /* noop */
+    } else if (
+      field.type instanceof GraphQLNonNull &&
+      field.type.ofType instanceof GraphQLScalarType
+    ) {
+      field.type = new GraphQLNonNull(new UpperCaseType(field.type.ofType));
+    } else if (field.type instanceof GraphQLScalarType) {
+      field.type = new UpperCaseType(field.type);
```

Comment:
Oh, looks like it's not a scalar but a custom transform trick 😕

Reply:
@19majkel94 yeah, again, it's how the `apollo` documentation does it.

Reply:
@19majkel94 let me know if you think there's another way of doing so.  I went around to a few repositories and others do it like this as well.

Reply:
I have not enough experience with directives as I don't use them. It's ok for me if it's a standard thing.